File: App/src/main/java/com/drownedinsound/ui/post/BoardPostPresenter.java
package com.drownedinsound.ui.post;


import com.drownedinsound.data.DisBoardRepo;
import com.drownedinsound.data.generatered.BoardPost;
import com.drownedinsound.data.generatered.BoardPostSummary;
import com.drownedinsound.ui.base.Navigator;

import android.support.annotation.NonNull;

import rx.Observer;
import rx.Scheduler;
import rx.Subscription;
import rx.functions.Func2;

import static rx.Observable.zip;

public class BoardPostPresenter implements BoardPostContract.Presenter {

    private final String boardPostId;

    private final String boardListType;

    private final BoardPostContract.View boardPostView;

    private final Navigator navigator;

    private final DisBoardRepo disBoardRepo;

    private final Scheduler mainThreadScheduler;

    private final Scheduler backgroundThreadScheduler;

    private Subscription loadBoardPostSubscription;

    public BoardPostPresenter(
            String postID,
            String boardListType,
            BoardPostContract.View boardPostView,
            Navigator navigator,
            Scheduler mainThreadScheduler,
            Scheduler backgroundThreadScheduler,
            DisBoardRepo disBoardRepo) {
        this.boardPostId = postID;
        this.boardListType = boardListType;
        this.boardPostView = boardPostView;
        this.navigator = navigator;
        this.disBoardRepo = disBoardRepo;
        this.mainThreadScheduler = mainThreadScheduler;
        this.backgroundThreadScheduler = backgroundThreadScheduler;
    }


    @Override
    public void onViewCreated() {
        loadBoardPost(false);
    }

    private void loadBoardPost(boolean forceUpdate) {
        boardPostView.showLoadingProgress(true);
        loadBoardPostSubscription = zip(
                disBoardRepo.getBoardPost(boardListType, boardPostId, forceUpdate),
                disBoardRepo.getBoardPostSummary(boardListType, boardPostId),
                getUpdateNumberOfTimesOpened())
                .subscribeOn(backgroundThreadScheduler)
                .observeOn(mainThreadScheduler)
                .subscribe(new Observer<BoardPost>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        boardPostView.showErrorView();
                        boardPostView.showLoadingProgress(false);
                    }

                    @Override
                    public void onNext(BoardPost boardPost) {
                        boardPostView.showBoardPost(boardPost);
                        boardPostView.showLoadingProgress(false);
                    }
                });
    }

    @NonNull
    private Func2<BoardPost, BoardPostSummary, BoardPost> getUpdateNumberOfTimesOpened() {
        return new Func2<BoardPost, BoardPostSummary, BoardPost>() {
            @Override
            public BoardPost call(BoardPost boardPost,
                    BoardPostSummary boardPostSummary) {
                if (boardPostSummary != null) {
                    boardPost.setNumberOfTimesOpened(
                            boardPostSummary.getNumberOfTimesOpened());
                }

                return boardPost;
            }
        };
    }

    @Override
    public void onViewDestroyed() {
        if(loadBoardPostSubscription != null) {
            loadBoardPostSubscription.unsubscribe();
        }
    }
}

File: App/src/main/kotlin/com/drownedinsound/ui/post/BoardPostPresenter.kt
package com.drownedinsound.ui.post

import com.drownedinsound.BoardPostItem
import com.drownedinsound.core.DisBoardsAppModule
import com.drownedinsound.data.DisBoardRepo
import com.drownedinsound.ui.base.Navigator
import rx.Scheduler
import rx.Subscription
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Named

class BoardPostPresenter @Inject constructor(
        @param:Named("postId") private val boardPostId: String,
        @param:Named("boardPostType") private val boardListType: String,
        private val boardPostView: BoardPostContract.View,
        private val navigator: Navigator,
        @param:Named(DisBoardsAppModule.MAIN_THREAD_SCHEDULER) private val mainThreadScheduler: Scheduler,
        @param:Named(DisBoardsAppModule.BACKGROUND_THREAD_SCHEDULER) private val backgroundThreadScheduler: Scheduler,
        private val disBoardRepo: DisBoardRepo,
        private val boardPostModelMapper: BoardPostModelMapper) : BoardPostContract.Presenter {

    private var loadBoardPostSubscription: Subscription? = null

    override fun onViewCreated() {
        boardPostView.setPresenter(this)
        loadBoardPost(false)
    }

    private fun loadBoardPost(forceUpdate: Boolean) {
        boardPostView.showLoadingProgress(true)
        loadBoardPostSubscription =
                disBoardRepo.getBoardPost(boardListType, boardPostId, forceUpdate)
                        .map(boardPostModelMapper::map)
                        .subscribeOn(backgroundThreadScheduler)
                        .observeOn(mainThreadScheduler)
                        .subscribe(this::displayItems, this::handleError)
    }

    private fun handleError(throwable: Throwable) {
        Timber.e(throwable,"Error getting board post")
        boardPostView.showErrorView()
        boardPostView.showLoadingProgress(false)
    }

    private fun displayItems(items: List<BoardPostItem>) {
        boardPostView.showBoardPostItems(items)
        boardPostView.showLoadingProgress(false)
    }

    override fun onViewDestroyed() {
        loadBoardPostSubscription?.unsubscribe()
    }

    override fun handleBackAction() = navigator.hideCurrentScreen()

    override fun handleRefreshAction() {
        if(loadBoardPostSubscription ==  null
                || loadBoardPostSubscription!!.isUnsubscribed) {
            loadBoardPost(true)
        }
    }
}